packages/dds/tree/src/core/tree/anchorSet.ts

```diff
@@ -131,6 +131,30 @@ export interface AnchorEvents {
 	 */
 	subtreeChanging(anchor: AnchorNode): PathVisitor | void;
 
+	/**
+	 * Emitted after the subtree rooted at `anchor` may have been changed.
+	 * 
+	 * @remarks
+	 * While this event is always emitted in the presence of changes to the subtree,
+	 * it may also be emitted even though no changes have been made to the subtree.
+	 * May be emitted multiple times within the application of a single edit or transaction.
+	 *
+	 * @privateremarks
+	 * The implementation details of the delta visit traversal have a big effect on how/when this event is emitted.
+	 * As of 2024-03-27, it is emitted on a node as the delta traversal exits that node, which by design of the current
+	 * traversal algorithm happens after all changes to the node's children have been applied, so this event is only
+	 * emitted by a node after all its children which had changes have emitted theirs.
+	 * It's also emitted twice for any node due to the two-pass algorithm (detach + attach) we use today which causes
+	 * a node to be exited once on each pass, and depending on the kind of change, the change may or may not have happened
+	 * yet the first time the event fires. E.g. replacing a node with a new one only produces a visible change during the
+	 * attach pass, so when this event fires during the detach pass, if the listener looks for a change in the tree, it
+	 * will not find one.
```

Comment:
I think this is misleading: "[this event] is emitted on a node as the delta traversal exits that node, which [...] happens after all changes to the node's children have been applied, so this event is only emitted by a node after all its children which had changes have emitted theirs". For example, if node A is the parent of node B, we'll emit an event for a detach under A before we emit an event for an attach under B.

This whole paragraph is much too detailed. The fact that it is hard to write it accurately is a symptom that doing so amounts to conveying in English the details of the delta visit algorithm. I'd like to understand what goal you're trying to meet by providing this information. I don't think there's any information in this paragraph that couldn't become outdated at some later date, so I wouldn't want anyone to depend on it (and I'd rather not have to remember to update this comment when we change the delta visit code).

Reply:
> For example, if node A is the parent of node B, we'll emit an event for a detach under A before we emit an event for an attach under B.

Is that correct? Wouldn't the detach pass still traverse B (enter, realize there's nothing to do, exit) and fire `subtreeChanged` in B _even if nothing happened in it during the detach pass_, then eventually exit A and fire subtreeChanged there?

Reply:
The motivation here is basically just helping ourselves (especially people new to this part of the codebase, like me 😄) get a better understanding for the "weirdness" of the API, like why it may fire when no changes are visible and/or may fire several times. I would hope the timestamp makes it clear that those remarks could be out of date and should be taken with a grain of salt. That said, I could also see that paragraph be just PR discussion, so I'm ok removing it if we think it'll be more of a maintenance issue than it is helpful. Will see if any other reviewers have any thoughts.

Reply:
> Wouldn't the detach pass still traverse B (enter, realize there's nothing to do, exit) and fire subtreeChanged in B even if nothing happened in it during the detach pass, then eventually exit A and fire subtreeChanged there?

My bad, you're correct. I've internalized the "lazy descent" model that I mentioned when we last chatted, but that hasn't been implemented yet. So, the current description is accurate until we remedy the unnecessary visits.

I agree the timestamp helps lower expectations as to the accuracy of the comment. As to whether it's best to have a comment whose accuracy cannot be trusted, or not to have that comment, I'm not sure. I think that I'll be sad either way until the weirdness and the comment are both gone, so if the comment can help onboarders get a foothold in the meantime, I'm okay with it. Feel free to resolve this comment thread without further action.

Reply:
Since both of you agree on this maybe being too much for this particular code location, I'm happy to move it to a better place in the delta visit code and leave a reference like Noah suggests. Will push that tomorrow morning :).

Reply:
FWIW, the one thing we could say that isn't likely to change and might help explain some of the emission patterns for this event is that if the event is emitted at least once for a node then a similar event will later be emitted at least once for its ancestors.